### ISSUE TITLE
List Text Element

### ISSUE THREAD

**Issue by LittleTal** [wix/wix-style-react#4048]
# ✨ Feature Request

### 🔦 Scope

Change default behaviour for `<li>` inside `<Text>`
Need to replace default html list style with `<Confirm/>` icon

![image](https://user-images.githubusercontent.com/23282301/64185370-b41ff700-ce75-11e9-8786-5d549ab6e2ca.png)

...

### 💁 Explanation

Currently when adding `<li>` to `<Text>` the result will be:
![image](https://user-images.githubusercontent.com/23282301/64185051-25ab7580-ce75-11e9-8159-916d5fe72278.png)

Need to change it to the style below:

![image](https://user-images.githubusercontent.com/23282301/64185108-3b209f80-ce75-11e9-95d1-ad7ef7686656.png)

**Required Styles**
Margin-right:6px
![image](https://user-images.githubusercontent.com/23282301/64186189-162d2c00-ce77-11e9-8e0b-d64ff7faedd8.png)

...

### 🐾 Possible solution 

...

### 👀 Severity

Low

**Comment by argshook** [comment]
it should be `<li>` that contains `<Text>`, not the other way around. as in:

bad:

```jsx
<Text>
  <li>
    <Box>
      <Box marginRight="6px">
        <Confirm/>
      </Box>
      Hello text
    </Box>
  </li>
</Text>
```

good:

```jsx
<li>
  <Box>
    <Box marginRight="6px">
      <Confirm/>
    </Box>

    <Text>Hello text</Text>
  </Box>
</li>
```

Is this the case? if not, maybe you have a broader code example?

**Comment by shlomitc** [comment]
@argshook - I think what Lital is trying to define is the default `li` behavior. Whenever the `li` element appears, a "V" mark should be displayed instead of the default browser styles (the `Box` example is just for the design POC).

**Comment by argshook** [comment]
ah, i see, thanks for clarifying.

it's possible to do it for native `li` through some hackish css, but might be problematic. What if someone wants not `<Confirm/>` icon but something else, say `<X/>`.

giving more thought to this, functionality would probably grow to constitute dedicated component, like, for example:

```
import { Ul, Li } from 'wix-style-react/List';

<Ul>
  <Li>by default Confirm</Li>
  <Li icon={<X/>}>possible to change</Li>
  <Li disabled>other props</Li>
</Ul>
```

**Comment by shlomitc** [comment]
@LittleTal - WDYT?

**Comment by LittleTal** [comment]
@argshook dedicated components sounds good. As discussed with @shlomitc, the structure should be:
```
<Text>
    <Ul>
        <Li>
        </Li>
    </Ul>
</Text>
```
(`<Text>` style should affect `<Ul>` and `<Li>` components)

**Comment by wuwa** [comment]
+1 this please, and IMO no need for more components, this is an overkill.
all is needed is CSS for regular ul inside a Text:
ul {
  list-style-image: url('icon.svg');
  padding: 0;
  margin: 0;
  list-style-position: inside;
}

**Comment by simast** [comment]
Hey, I am working on this issue right now, but the way I see it we have two options:

1. Just update styling to all lists inside `<Text>...</Text>` component as suggested by @wuwa. The drawback is that users will not be able to change list icon and this will be a potentially breaking change for any existing users that already use `<ul>`/`<li>` elements in a text component (potentially with their own styling).
2. Create completely new components as suggested by @argshook. The benefits here are then we can provide more options to the users - like the ability to change the icon. We will also not break any existing layouts since this will be a new component. If we are going this route should we create specs for this component? Like we need to define proper spacing and other details (such as whether you can use a list inside another list like you can with the default `<ul>`/`<li>` elements).

I personally think we should create separate components (second option). Looking for more feedback before I will implement this.

**Comment by simast** [comment]
Some other issues I see with all the suggestions above:

```jsx
<Text ellipsis>
  <Ul>
    <Li>Test</Li>
  </Ul>
</Text>
```

...this will not work as expected.

Also this one:

```jsx
<Text size="tiny">
  <Ul>
    <Li>Test</Li>
  </Ul>
</Text>
```

... will have tiny text but rather large icons (we should probably use `ConfirmSmall` icon in this case and not `Confirm`).

What if we implement this as a separate component but with support with the same properties as a `<Text>` component (except some that do not work as expected like `ellipsis`)? So you could do:

```jsx
<Ul size="tiny" skin="premium">
  <Li>Test</Li>
</Ul>
```

This doesn't mean we will have to duplicate props and styles - we will use Stylable support to extend `<Text>` component (I have a proof of concept implemented for this working as expected).

**Comment by argshook** [comment]
> I have a proof of concept implemented using this approach at the moment

well then, push a PR, mister! :)

I also tend to think that even though dedicated `Ul`/`Li` (or similarly
named) components might be overkill, it's likely the most viable and
future-proof option we have

**Comment by simast** [comment]
Storybook with some examples for the POC: https://wix-wix-style-react-pr-4531.surge.sh/?selectedKind=Components&selectedStory=List&full=0&addons=0&stories=1&panelRight=0

**Comment by simast** [comment]
After some internal discussion I propose the following API:

The component would be called "TextList" (to suggest that you can only build a list of text and not with other items/nodes).

```jsx
import TextList from 'wix-style-react/TextList';

// for unordered lists
<TextList icon="..." size="small" skin="premium">
  <TextList.Item>First item</TextList.Item>
  <TextList.Item>Second item</TextList.Item>
  <TextList.Item>Third item</TextList.Item>
</TextList>

// for ordered lists
<TextList as="ol" size="small" skin="premium">
  <TextList.Item>First item</TextList.Item>
  <TextList.Item>Second item</TextList.Item>
  <TextList.Item>Third item</TextList.Item>
</TextList>
```

- We will use compound `Item` component to define `<li>` - as a hint that this component should not be used outside of `TextList` and for consistency with other parts of the library.
- All props from `<Text>` will be supported (like skin, weight, size, light, secondary) - except those that do not make sense, like `ellipsis`.

**Comment by argshook** [comment]
> The component would be called "TextList" (to suggest that you can only build a list of text and not with other items/nodes).

`PropType.node` should be still supported, because what if user wants `TextButton` or `Tooltip`. Those cases are valid imo.

**Comment by simast** [comment]
> `PropType.node` should be still supported, because what if user wants `TextButton` or `Tooltip`. Those cases are valid imo.

Yes, of course! You should also be able to nest `<TextList>` inside `<TextList.Item>` to create nested list (just like you can with regular `<ul>`/`<li>` tag nesting). I will not add any explicit PropType restrictions to children.

**Comment by simast** [comment]
@LittleTal should we define specs for this component then? Some questions that I am not sure about yet and would like to have in a spec:

- What's the default spacing around text list component (if there is one).
- What's the spacing for nested text list components.
- What's the default icon for unordered text list (and if the icon should change if you have a nested list).
- What are the other states (should be similar to text component).
- Do we want to allow users specifying any custom icon, just a set of valid icons or no custom icons at all?

**Comment by LittleTal** [comment]
@wuwa WDYT?

**Comment by wuwa** [comment]
i still think this is an overkill.

1. currently the default ul/li styling is not good - nobody is using it (or should). i believe most lists currently are built without html list tags but with custom spans / boxes.
2. the point of defining this in DS level is forcing a unified look for basic list (no custom icons). if various icons needed for specific lists, anyway this can be achieved currently using icons and boxes.
3. the most valid usecase of such custom icons is actually different icons inside the same list, which is not solved.
4. clean markup is also something to consider, all these extra components are over-complex imo.

to summarize - i think default list styling should be fixed. if consumers need a simple list this will speed up their dev time. if need their own lists they can build it today already quite easily.

**Comment by simast** [comment]
Some answers to @wuwa comments:

1. In `wix-style-react` we try not to style HTML elements directly - for example `<h1>` or `<p>` has no default styles provided by the library, so I do not understand why we should treat `<ul>`/`<li>` differently here. Also `<Text>` was designed for inline text, so having block elements inside complicates things.
2. We don't want users to create custom lists using icons and boxes - as that would be inconsistent and users will loose accessibility features provided by <ul>/<li> (which we will use in our `TextList` component internally).
3. No problem, we can add `<TextList.Item icon="...">` if this feature is needed.
4. I agree that it does add a bit of a noise, but I do think it's a better approach and would scale better for any future needs.

**Comment by mykas** [comment]
@wuwa If we add block based components support in Text component - we will have a broken ellipsis support on that. We can't handle `ul` and `li` inside the Text with ellipsis support.
